### PR TITLE
Multi byte truncate

### LINKED ISSUE #414: Calling a customized entry_maker causes an error
### Description

I'm developing [telescope-z.nvim][]. I implemented [entry_maker][] with referring to [gen_from_git_commits][]. My implementation has almost the same logic as `gen_from_git_commits`'s one, but Neovim raises an error with mine.

[entry_maker]: https://github.com/delphinus/telescope-z.nvim/blob/45bdd99d2315f35a151bd303a66b81beb794df45/lua/telescope/_extensions/z_builtin.lua#L11-L38
[gen_from_git_commits]: https://github.com/nvim-telescope/telescope.nvim/blob/9503603f888305ec799c9346f51a6e3da7f5b89b/lua/telescope/make_entry.lua#L262-L292
[telescope-z.nvim]: https://github.com/delphinus/telescope-z.nvim

**Expected Behavior**

`:Telescope z list` causes no error.

**Actual Behavior**

`:Telescope z list` causes an error below.

```vim
Error executing luv callback:
...t/telescope.nvim/lua/telescope/pickers/entry_display.lua:6: E5560: vimL function must not be called in a lua loop callback
Error executing luv callback:
cannot resume dead coroutine
```

`entry_display.lua:6` is calling `vim.fn.strdisplaywidth()`. I comment out this `if` block, then no error occurs.

https://github.com/nvim-telescope/telescope.nvim/blob/de80a9837cd1d207981c1f6dbf504436f8bfee13/lua/telescope/pickers/entry_display.lua#L6

### Details

<details><summary>Reproduce</summary>

1. nvim test.lua

   ```lua
   local install_path = vim.fn.stdpath'data'..'/site/pack/packer/opt/packer.nvim'
   if vim.fn.isdirectory(install_path) == 0 then
     os.execute('git clone https://github.com/wbthomason/packer.nvim '..install_path)
     vim.cmd'packadd packer.nvim'
   end

   require'packer'.startup(function()
     use'wbthomason/packer.nvim'
     use{
       'nvim-telescope/telescope.nvim',
       requires = {
         'nvim-lua/popup.nvim',
         'nvim-lua/plenary.nvim',
         {'delphinus/telescope-z.nvim', branch = 'feature/entry-maker'},
       },
       config = function()
         vim.cmd[[packadd telescope-z.nvim]]
         require'telescope'.load_extension'z'
       end,
     }
   end)
   ```

2. nvim -i NONE -nu test.lua
3. `:PackerInstall` then `:PackerCompile` then `:qa`
4. nvim -i NONE -nu test.lua
5. `:Telescope z list`
6. An error occurs.

   ```vim
   Error executing luv callback:
   ...t/telescope.nvim/lua/telescope/pickers/entry_display.lua:6: E5560: vimL function must not be called in a lua loop callback
   Error executing luv callback:
   cannot resume dead coroutine
   ```
</details>

<details><summary>Environment</summary>

- nvim version: NVIM v0.5.0-dev+1005-ga1ed941a7
- Operating system: macOS 10.15.7
- Telescope commit: de80a9837cd1d207981c1f6dbf504436f8bfee13

</details>

<details><summary>Configuration</summary>
<p>
 

```lua
local install_path = vim.fn.stdpath'data'..'/site/pack/packer/opt/packer.nvim'
if vim.fn.isdirectory(install_path) == 0 then
  os.execute('git clone https://github.com/wbthomason/packer.nvim '..install_path)
  vim.cmd'packadd packer.nvim'
end

require'packer'.startup(function()
  use'wbthomason/packer.nvim'
  use{
    'nvim-telescope/telescope.nvim',
    requires = {
      'nvim-lua/popup.nvim',
      'nvim-lua/plenary.nvim',
      {'delphinus/telescope-z.nvim', branch = 'feature/entry-maker'},
    },
    config = function()
      vim.cmd[[packadd telescope-z.nvim]]
      require'telescope'.load_extension'z'
    end,
  }
end)
```

### PR DESCRIPTION
**Pull request by delphinus**
Fix #414
Ref #413
Ref #458
Ref https://github.com/nvim-telescope/telescope.nvim/pull/274#discussion_r528363826

`entry_display.truncate()` has 2 problems.

1. This func will be executed in Lua loop, but Neovim cannot use `vim.fn.foo()` in it.
2. It breaks UTF-8 bytes row when trying to truncate multi-byte characters.

I solved them by implementing 2 VimL functions in Lua: `strdisplaywidth()`, `strcharpart()`. I simply convert C logic from Neovim source. ([here][1] and [here][2])

[1]: https://github.com/neovim/neovim/blob/a1ed941a7881122fda2fd48e71e890ed55e4d08e/src/nvim/eval/funcs.c#L9845-L9858
[2]: https://github.com/neovim/neovim/blob/a1ed941a7881122fda2fd48e71e890ed55e4d08e/src/nvim/eval/funcs.c#L9870-L9922

### CODE REVIEW

File: lua/telescope/pickers/entry_display.lua [hunk 1]
```diff
@@ -1,15 +1,27 @@
...
+      break
+    end
+    result = part
+    charlen = charlen + 1
   end
-  return str
+  return result
 end
 
+entry_display.truncate = truncate
```
Comment (Conni2461) [inline lua/telescope/pickers/entry_display.lua:23]: I like to have `entry_display.create` immediately rather then first creating a `local function`

File: lua/telescope/pickers/entry_display.lua [hunk 2]
```diff
@@ -1,15 +1,27 @@
...
+  end
+  local charlen = 0
+  local result = ''
+  while true do
+    local part = utils.strcharpart(str, 0, charlen) ..  '…'
+    if utils.strdisplaywidth(part) > len then
+      break
+    end
+    result = part
+    charlen = charlen + 1
```
Comment (Conni2461) [inline lua/telescope/pickers/entry_display.lua:18]: Hmmm isn't it not better to only look at the new part step by step and memorize it the prev result? I am just a little bit worried about performance :) I rewrote it to that: Any thoughts? Tests still pass
```suggestion
  local charlen = 0
  local cur_len = 0
  local result = ''
  local len_of_dots = utils.strdisplaywidth('…')
  while true do
    local part = utils.strcharpart(str, charlen, 1)
    cur_len = cur_len + utils.strdisplaywidth(part)
    if (cur_len + len_of_dots) > len then
      result = result .. '…'
      break
    end
    result = result .. part
    charlen = charlen + 1
  end
```

File: lua/tests/automated/pickers/entry_display_spec.lua [hunk 3]
```diff
@@ -0,0 +1,32 @@
+local entry_display = require('telescope.pickers.entry_display')
+
```
Comment (Conni2461) [inline lua/tests/automated/pickers/entry_display_spec.lua:2]: That file is in the wrong dir. Everything inside pickers are integration tests. Put it in `lua/tests/automated/`
But thank you very much for writing tests :)

**Comment by Conni2461** [comment]
I still need to look at `utils.strcharpart` i am doing that later today. Thanks for your work :)

Note: failing tests is unrelated. Will be fixed here https://github.com/nvim-lua/plenary.nvim/pull/57

File: lua/telescope/pickers/entry_display.lua (hunk 1 above)
Comment (delphinus) [reply lua/telescope/pickers/entry_display.lua]: Sure. fixed on 19188ce

File: lua/telescope/pickers/entry_display.lua (hunk 2 above)
Comment (delphinus) [reply lua/telescope/pickers/entry_display.lua]: Look good! committed on 1f77e33

File: lua/tests/automated/pickers/entry_display_spec.lua (hunk 3 above)
Comment (delphinus) [reply lua/tests/automated/pickers/entry_display_spec.lua]: Fixed on 45be77c

**Comment by tjdevries** [comment]
Is this OK to merge now? Looks great!

**Review by Conni2461 — APPROVED** [review]
Okay i looked at the last part. Sorry that took that long :)
LGTM
Thanks :)

**Pull request merged**